### DISCUSSION TITLE
Two installations with multiple sites each and same webroot

### DISCUSSION THREAD

**Discussion by jannisborgers** [craftcms/cms#12563] [Q&A] [answered]
Hi there,

a multi-site installation that was built a couple of years back currently has four sites:

* `deutsch.example.com/2020` (German 2020)
* `english.example.com/2020` (English 2020)
* `deutsch.example.com/2021` (German 2021)
* `english.example.com/2021` (English 2021)

This was solved with a multi-site setup and a custom module, because the root domain should always point to the latest year-segment (`deutsch.example.com` currently redirects to `deutsch.example.com/2021` etc.).

The second year’s content structure has already diverged noticeably (because editorial exceptions), so we want to put the upcoming year (2022) in a separate installation on the server. Because of the domain setup, we only have two webroots (one per language). 

We need two new site URLs, this time pointing to the new installation: 
* `deutsch.example.com/2022` (German 2022)
* `english.example.com/2022` (English 2022)

# Possible solution

I found this thread, which helped me identify that `Option 2` described by Brandon might work in our case:
https://craftcms.stackexchange.com/questions/37296/multiple-instances-of-craft-in-same-hosting-account-root

I set it up locally and now have three working directories:
* `web`, which holds the contents of the old `web` subfolder and which both local domains point to
* `old` with the old installation (sans `web` folder)
* `new` with a new installation (sans `web` folder)

I then modified the `index.php` inside `web` as follows:

```
…

// Set path constants
$uri = $_SERVER['REQUEST_URI'];
$uri_path = parse_url($uri, PHP_URL_PATH);
$uri_segments = explode('/', $uri_path);
$site_folder = ($uri_segments[1] == "2022" || strpos($uri, "new-admin") !== false) ? 'new' : 'old';
define('CRAFT_BASE_PATH', dirname(__DIR__) . "/" . $site_folder);
define('CRAFT_VENDOR_PATH', CRAFT_BASE_PATH . '/vendor');

…
```
I also set up a new database and fed its details into the new installation’s `.env` file.

# Questions

1. How can I access each installation’s control panel individually? 
    * ~~Currently when I visit `domain.test/admin` locally, I get a weird mix where I land in the old installation (with the four old sites), but the system name in the top left is taken from the new installation’s `.env`~~ (my bad, had the `.env` settings backwards)
    * ~~I have tried setting different `cpTrigger` settings in each installation’s `general.php` -> `old-admin` and `new-admin` respectively. I can now access `domain.test/old-admin` and get the old control panel, but `domain.test/new-admin` gives me an `Template not found: new-admin` error.~~
3. File uploads are now in one location for both installations. Can this be a problem when uploading files or modifying uploaded files from any of the two installations?
5. … anything else I should consider? I already found [this knowledge base article](https://craftcms.com/knowledge-base/running-multiple-craft-installations-from-the-same-domain), which I will follow as well, and of course this means a new license for the new installation.

# Update 1

I have updated question 1 with corrected information and tried setting different `cpTrigger` values.

# Update 2

Slowly getting there… and documenting for others to find.

* `/new-admin` URLs where referencing the old craft base path and thus throwing `Template not found` errors. To make sure that the old installation was responsible for these URLs, I had to make sure that URLs with that segment are correctly routed to the right `CRAFT_BASE_PATH` in `index.php`. Otherwise the old installation is used via the `CRAFT_BASE_PATH`, where `new-admin` is not the right `cpTrigger`.
* I also defined the `@web` and `@webroot` aliases specifically in both installations.

**Comment by brandonkelly** [comment, marked as answer]
Sounds like you’ve mostly got a handle on this.

> File uploads are now in one location for both installations. Can this be a problem when uploading files or modifying uploaded files from any of the two installations?

I would not recommend sharing the same upload folder between two installations. Otherwise, if you make any changes to a file’s location in one install (rename it or move it), it will simply go missing in the other install. And if you edit or replace the file, its new `dateModified` and dimension data will be out of sync.

**Reply by jannisborgers** [reply]
Thanks @brandonkelly. I can confirm that the approach works right now, I have moved it to the client server as well, without problems.

I'll set up new upload locations for the new installations, that was the last item on my list.
